src/profile-logic/profile-data.js

```diff
@@ -639,19 +639,48 @@ export function getTimingsForCallNodeIndex(
   return { forPath: pathTimings, forFunc: funcTimings, rootTime };
 }
 
+// This function computes the time range for a thread, using both its samples
+// and markers data.
 export function getTimeRangeForThread(
-  thread: Thread,
+  { samples, markers }: Thread,
   interval: Milliseconds
 ): StartEndRange {
-  if (thread.samples.length === 0) {
-    return { start: Infinity, end: -Infinity };
+  const result = { start: Infinity, end: -Infinity };
+
+  if (samples.length) {
+    const lastSampleIndex = samples.length - 1;
+    result.start = samples.time[0];
+    result.end = samples.time[lastSampleIndex] + interval;
   }
 
-  const lastSampleIndex = thread.samples.length - 1;
-  return {
-    start: thread.samples.time[0],
-    end: thread.samples.time[lastSampleIndex] + interval,
-  };
+  if (markers.length) {
+    // Finding start and end times sadly requires looping through all markers :(
+    let startTime = +Infinity;
+    let endTime = -Infinity;
+    for (let i = 0; i < markers.length; i++) {
```

Comment:
Could we sneak this check into some other selector that loops over all the markers? That way we'd only incur the loop cost once. Like something inside of:

```
   * 1. _getRawMarkerTable - Get the RawMarkerTable from the current thread.
   * 2. getProcessedRawMarkerTable - Process marker payloads out of raw strings, and
   *                                 other future processing needs. This returns a
   *                                 RawMarkerTable still.
   * 3a. _getDerivedMarkers        - Match up start/end markers, and start
   *                                 returning the Marker[] type.
   * 3b. _getDerivedJankMarkers    - Jank markers come from our samples data, and
   *                                 this selector returns Marker structures out of
   *                                 the samples structure.
   * 4. getFullMarkerList          - Concatenates and sorts all markers coming from
   *                                 different origin structures.
   * 5. getFullMarkerListIndexes   - From the full marker list, generates an array
   *                                 containing the sequence of indexes for all markers.
   * 5. getCommittedRangeFilteredMarkerIndexes - Apply the committed range.
   * 6. getSearchFilteredMarkerIndexes         - Apply the search string
   * 7. getPreviewFilteredMarkerIndexes        - Apply the preview range
```
We could compose the selectors to still maintain the existing APIs.

e.g.
```
getTimeRange => state => getDerivedMarkersAndMarkerTimeRange(state).timeRange
getDerivedMarkers => state => getDerivedMarkersAndMarkerTimeRange(state).markers
```

Reply:
Note that we already ran the loop once at load time. Indeed this is run on all threads in the reducer `rootRange` when we get the action `PROFILE_LOADED`... That's why I suggested to memoize at the function level.